Xamarin.Forms.Platform.iOS/Renderers/CarouselViewRenderer.cs

```diff
@@ -0,0 +1,884 @@
+			// then they contain the index where the item was replaced.
+			if (e.Action == NotifyCollectionChangedAction.Replace)
+			{
+				if (Element == null && _pageController == null && Source == null) return;
+
+				// Remove controller from ChildViewControllers
+				if (ChildViewControllers != null)
+					ChildViewControllers.RemoveAll(c => c.Tag == Source[e.OldStartingIndex]);
+
+				Source[e.OldStartingIndex] = e.NewItems[e.NewStartingIndex];
+
+				var firstViewController = CreateViewController(Element.Position);
+
+				_pageController.SetViewControllers(new[] { firstViewController }, UIPageViewControllerNavigationDirection.Forward, false, s =>
+				{
+					SetArrowsVisibility();
+					SetIndicatorsCurrentPage();
+				});
+			}
+
+			// No other properties are valid.
+			if (e.Action == NotifyCollectionChangedAction.Reset)
+			{
+				if (Element == null) return;
+
+				UpdateItemsSource();
+			}
+		}
+
+		View GetItem(int index)
+		{
+			return TemplatedItemsView.TemplatedItems[index];
+		}
+
+		int GetItemCount()
+		{
+			if (TemplatedItemsView == null)
+				return -1;
+
+			return TemplatedItemsView.TemplatedItems.Count;
+		}
+
+		void Scroller_Scrolled(object sender, EventArgs e)
+		{
+			var scrollView = (UIScrollView)sender;
+			var point = scrollView.ContentOffset;
+
+			double currentPercentCompleted;
+			ScrollDirection direction;
+
+			if (Element.Orientation == CarouselViewOrientation.Horizontal)
+			{
+				currentPercentCompleted = Math.Floor((Math.Abs(point.X - _pageController.View.Frame.Size.Width) / _pageController.View.Frame.Size.Width) * 100);
+				direction = _prevPoint > point.X ? ScrollDirection.Left : ScrollDirection.Right;
+				_prevPoint = point.X;
+			}
+			else
+			{
+				currentPercentCompleted = Math.Floor((Math.Abs(point.Y - _pageController.View.Frame.Size.Height) / _pageController.View.Frame.Size.Height) * 100);
+				direction = _prevPoint > point.Y ? ScrollDirection.Up : ScrollDirection.Down;
+				_prevPoint = point.Y;
+			}
+
+			if (currentPercentCompleted <= 100 && currentPercentCompleted > _percentCompleted)
+			{
+				Element.SendScrolled(currentPercentCompleted, direction);
+				_percentCompleted = currentPercentCompleted;
+			}
+			else
+			{
+				_percentCompleted = 0;
+			}
+		}
+
+		void PageController_DidFinishAnimating(object sender, UIPageViewFinishedAnimationEventArgs e)
+		{
+			if (e.Completed)
+			{
+				var controller = (FormsUIViewContainer)_pageController.ViewControllers[0];
+				var position = Source.IndexOf(controller.Tag);
+				_isChangingPosition = true;
+				UpdateRendererPosition(position);
+				_isChangingPosition = false;
+				SetArrowsVisibility();
+				SetIndicatorsCurrentPage();
+			}
+		}
+
+		void SetIsSwipeEnabled()
+		{
+			foreach (var view in _pageController?.View.Subviews)
+			{
+				if (view is UIScrollView scroller)
+				{
+					scroller.ScrollEnabled = Element.IsSwipeEnabled;
+					scroller.Scrolled += Scroller_Scrolled;
```

Comment:
This event is never unsubscribed, this can cause a NullReferenceException when the Element is accessed in the handler method (Scroller_Scrolled).
We have crash reports of that.
And if the event is called after the dispose it can be also a source of memory leak.